### PR TITLE
Fix DeepSpeed SP loss aggregation and LocalLayerRepository kwargs

### LINKED ISSUE #47068: Potential host synchronization in deepspeed_sp_compute_loss due to Python conditional on CUDA tensor
### System Info

- Transformers: 4.57.2
- PyTorch: 2.7.1+cu128
- DeepSpeed: 0.17.6
- CUDA: 12.8
- GPU: NVIDIA B300 SXM6 (2 nodes × 8 GPUs, 16 GPUs total)
- Interconnect:
  - Intra-node: NVSwitch
  - Inter-node: InfiniBand

### Information

While profiling a distributed training workload with Nsight Systems, I observed what appears to be repeated host synchronization inside `deepspeed_sp_compute_loss`. The relevant code is located in: `src/transformers/integrations/deepspeed.py`

```python
good_tokens_per_rank = torch.distributed.nn.functional.all_gather(good_tokens, group=sp_group)

total_loss = sum(
    losses_per_rank[rank] * good_tokens_per_rank[rank]
    for rank in range(sp_world_size)
    if good_tokens_per_rank[rank] > 0
)
```

Assuming `good_tokens_per_rank` contains CUDA tensors, the Python conditional appears to evaluate a CUDA tensor in Python, which may introduce an implicit host synchronization.

https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/deepspeed.py
```python
if good_tokens_per_rank[rank] > 0
```

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Possible implementation

Would it make sense to avoid the Python conditional and perform the filtering entirely on the GPU?
For example, something along the lines of:

```python
good_tokens = torch.stack(good_tokens_per_rank)
losses = torch.stack(losses_per_rank)

mask = good_tokens > 0

# Preserve the original "skip zero-token ranks" behavior.
safe_losses = torch.where(
    mask,
    losses,
    torch.zeros_like(losses),
)

total_loss = (safe_losses * good_tokens).sum()
```

This could potentially avoid evaluating CUDA tensors in Python while preserving the same semantics. However, I'm not sure whether the current implementation intentionally skips zero-token ranks to avoid NaN propagation or for some other reason, so I'd appreciate your thoughts before attempting a PR.

### Expected behavior

The loss aggregation should avoid unnecessary host synchronization while preserving the current semantics. If this behavior is intentional, it would be helpful to understand the rationale. Otherwise, I'd be happy to submit a PR after discussing the preferred implementation.

Thank you!

### PR DESCRIPTION
**Pull request by sshivampeta**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47073)

## What does this PR do?
Fixes weighted loss aggregation in DeepSpeed sequence parallelism and unblocks make check-repo by aligning LocalLayerRepository calls with the installed kernels API.

DeepSpeed SP loss (#47068): deepspeed_sp_compute_loss used a Python if good_tokens_per_rank[rank] > 0 loop over gathered tensors. That forces a CUDA host sync on every step and lets NaN losses from zero-token ranks (e.g. SFT ranks with only masked prompt tokens) pollute the aggregated loss. This PR replaces that loop with pure tensor ops via a small _aggregate_weighted_sp_loss helper using torch.where, preserving the same weighted average while staying on-device and ignoring zero-token ranks safely.

Kernels build fix: kernel_config.py and hub_kernels.py passed an unsupported package_name kwarg to LocalLayerRepository. The installed kernels package (0.15.x / 0.16.x) only accepts repo_path and layer_name, which caused ty to fail during make check-repo and would raise TypeError at runtime for local kernel loading. Removed the invalid argument to match the upstream API.

Fixes #47068

Motivation
Reported in #47068: DeepSpeed sequence-parallel training can produce incorrect or NaN aggregated loss when some SP ranks have zero valid tokens. The root cause is the Python conditional over CUDA tensors in the loss aggregation path.
The LocalLayerRepository change is included because without it the repo does not pass type checking locally, blocking validation of this fix.
Changes
•Add _aggregate_weighted_sp_loss() in src/transformers/integrations/deepspeed.py
•Use it from deepspeed_sp_compute_loss() instead of the Python sum/conditional loop
•Remove invalid package_name from LocalLayerRepository calls in kernel_config.py and hub_kernels.py
•Add unit tests in tests/integrations/test_deepspeed_sp_loss.py

Test plan
[x] pytest tests/integrations/test_deepspeed_sp_loss.py -v — 5/5 passed
[x] make check-repo — 24/24 passed

AI assistance
AI tools were used to draft and iterate on this change. The submitter reviewed all changed lines and ran the tests above locally.
Code Agent Policy
[x] I confirm that this is not a pure code agent PR.
Before submitting
[ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
[x] Did you read the contributor guideline and the Pull Request checks?
[ ] Was this discussed/approved via a Github issue or the forum? Please add a link to it if that's the case.
[ ] Did you make sure to update the documentation with your changes according to the guidelines?
[x] Did you write any new necessary tests?

## Who can review?
@SunMarc @3outeille @vasqu

- AMD ROCm: @ivarflakstad
- Intel XPU: @IlyasMoutawwakil
- Ascend NPU: @ivarflakstad 
- Documentation: @stevhliu

### CODE REVIEW

**Comment by hdimmfh** [comment]
Thanks for working on this so quickly.

I originally reported this issue in #47068 after identifying the repeated 8-byte D2H memcpy and host-side `pthread_cond_wait` through Nsight Systems profiling. The GPU-side aggregation direction in this PR matches the approach I had in mind.

If you think my issue analysis and proposed direction warrant it, I’d be grateful if you would consider adding me as a co-author. If not, no worries — I appreciate you implementing the fix.

Co-authored-by: Myeongsu Koo <rnaudtndi@naver.com>

**Comment by sshivampeta** [comment]
Thanks for the detailed profiling in #47068 — the Nsight trace identifying the repeated 8-byte D2H memcpy and the host-side pthread_cond_wait was what made the root cause clear. Happy to add you as co-author; I'll update the commit with the Co-authored-by trailer.

**Comment by hdimmfh** [comment]
> Thanks for the detailed profiling in #47068 — the Nsight trace identifying the repeated 8-byte D2H memcpy and the host-side pthread_cond_wait was what made the root cause clear. Happy to add you as co-author; I'll update the commit with the Co-authored-by trailer.

Thanks a lot! I really appreciate the co-author credit. Glad the profiling was helpful!

**Comment by sshivampeta** [comment]
@SunMarc Could you review when you have a chance? This addresses the D2H sync in `deepspeed_sp_compute_loss` reported in #47068 (cc'd to you by @Rocketknight1). Happy to adjust based on feedback.

File: src/transformers/integrations/deepspeed.py [hunk 1]
```diff
@@ -688,6 +688,23 @@ def propagate_args_to_deepspeed(accelerator, args, auto_find_batch_size=False):
...
+    polluting the total (see https://github.com/huggingface/transformers/issues/47068).
+    """
+    losses_stacked = torch.stack(losses_per_rank)
+    good_tokens_stacked = torch.stack(good_tokens_per_rank)
+    mask = good_tokens_stacked > 0
+    safe_losses = torch.where(mask, losses_stacked, torch.zeros_like(losses_stacked))
+    total_loss = (safe_losses * good_tokens_stacked).sum()
+    total_good_tokens = good_tokens_stacked.sum()
+    return total_loss / total_good_tokens.clamp(min=1)
+
```
Comment (SunMarc) [inline src/transformers/integrations/deepspeed.py]: don't create a new function

File: tests/integrations/test_deepspeed_sp_loss.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import random
+import unittest
+
+import torch
```
Comment (SunMarc) [inline tests/integrations/test_deepspeed_sp_loss.py]: we don't need this. The existing tests should cover this

**Review by SunMarc — COMMENTED** [review]
Thanks, just some nits. Does this fixes your issue @hdimmfh ?

**Comment by hdimmfh** [comment]
> Thanks, just some nits. Does this fixes your issue @hdimmfh ?

Yes, this resolves the issue I reported. Thanks for reviewing!

**Comment by sshivampeta** [comment]
Addressed both comments @SunMarc :
•Inlined the weighted aggregation into deepspeed_sp_compute_loss (removed _aggregate_weighted_sp_loss).
•Removed tests/integrations/test_deepspeed_sp_loss.py per your note that existing tests should cover this.
@hdimmfh — does this match what you saw in your Nsight profiling?

File: src/transformers/integrations/deepspeed.py (hunk 1 above)
Comment (sshivampeta) [reply src/transformers/integrations/deepspeed.py]: thanks for the review

File: tests/integrations/test_deepspeed_sp_loss.py (hunk 2 above)
Comment (sshivampeta) [reply tests/integrations/test_deepspeed_sp_loss.py]: thanks for the review

**Review by SunMarc — APPROVED** [review]
Thanks !

**Comment by sshivampeta** [comment]
Thanks for the review, @SunMarc! CI is green — ready to merge when you have a moment.

**Comment by hdimmfh** [comment]
Yes, this matches what I observed in my Nsight profiling. The changes still seem appropriate for addressing the issue. Thanks for addressing the review comments!
@sshivampeta

**Comment by hdimmfh** [comment]
This looks like a flaky CI issue (dist.barrier() → Gloo Connection reset by peer) rather than a regression from this PR. Could we rerun or re-enqueue it @SunMarc  ?

**Comment by sshivampeta** [comment]
@SunMarc All checks are green and the branch is mergeable — ready to merge whenever you have a moment. Thanks!
